services/insights-handler/internal/handler/main.go

```diff
@@ -0,0 +1,713 @@
+	}
+
+	//@todo replace this by checking if incoming.Payload and converting to gzip, or vise-versa, so we don't repeat here
+	if incoming.Payload != nil {
+		for _, v := range incoming.Payload {
+			if insights.InputType == Sbom {
+				err := h.processSbomInsightsData(insights, v, apiClient, resource)
+				if err != nil {
+					log.Println(fmt.Errorf(err.Error()))
+				}
+			}
+		}
+	}
+
+	if incoming.BinaryPayload != nil {
+		for _, v := range incoming.BinaryPayload {
+			if insights.InputType == SbomGz {
+				err := h.processSbomInsightsData(insights, v, apiClient, resource)
+				if err != nil {
+					log.Println(fmt.Errorf(err.Error()))
+				}
+			}
+
+			if insights.InputType == ImageInspectGz {
+				err = h.processImageInspectInsightsData(insights, v, apiClient, resource)
+				if err != nil {
+					log.Println(fmt.Errorf(err.Error()))
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+func (h *Messaging) processSbomInsightsData(insights InsightsData, v string, apiClient graphql.Client, resource ResourceDestination) error {
+	bom := new(cdx.BOM)
+
+	if insights.InputType == Sbom {
+		b := []byte(v)
+		decoder := cdx.NewBOMDecoder(bytes.NewReader(b), cdx.BOMFileFormatJSON)
+		if err := decoder.Decode(bom); err != nil {
+			return err
+		}
+	}
+
+	if insights.InputType == SbomGz {
+		result, err := decodeString(v)
+		if err != nil {
+			return err
+		}
+		b, err := json.MarshalIndent(result, "", " ")
+		if err != nil {
+			return err
+		}
+
+		decoder := cdx.NewBOMDecoder(bytes.NewReader(b), cdx.BOMFileFormatJSON)
+		if err = decoder.Decode(bom); err != nil {
+			panic(err)
+		}
+	}
+
+	project, environment, apiErr := determineResourceFromLagoonAPI(apiClient, resource)
+	if apiErr != nil {
+		return apiErr
+	}
+	source := fmt.Sprintf("%s:%s", (*bom.Metadata.Component).Name, resource.Service)
+
+	apiErr = h.deleteExistingFactsBySource(apiClient, environment, source, project)
+	if apiErr != nil {
+		return apiErr
+	}
+
+	// Process SBOM into facts
+	facts := processFactsFromSBOM(bom.Components, environment.Id, source)
+	log.Printf("Successfully decoded SBOM of image %s\n", bom.Metadata.Component.Name)
+	log.Printf("- Generated: %s with %s\n", bom.Metadata.Timestamp, (*bom.Metadata.Tools)[0].Name)
+	log.Printf("- Packages found: %d\n", len(*bom.Components))
+
+	apiErr = h.pushFactsToLagoonApi(facts, resource)
+	if apiErr != nil {
+		return apiErr
+	}
+	return nil
+}
+
+func (h *Messaging) processImageInspectInsightsData(insights InsightsData, v string, apiClient graphql.Client, resource ResourceDestination) error {
+	if insights.InputType == ImageInspectGz {
+		decoded, err := decodeString(v)
+		if err != nil {
+			fmt.Errorf(err.Error())
+		}
+
+		project, environment, apiErr := determineResourceFromLagoonAPI(apiClient, resource)
+		if apiErr != nil {
+			return apiErr
+		}
+		source := fmt.Sprintf("image-inspect:%s", resource.Service)
+		apiErr = h.deleteExistingFactsBySource(apiClient, environment, source, project)
```

Comment:
I'm wondering if this means we're deleting the existing facts before we know whether or not we can successfully unmarshall/process the new ones? Should just be just before we push the new facts?